src/s3access.cpp

```diff
@@ -110,37 +96,60 @@ QString S3Access::bucket() const
 
 S3ListResult *S3Access::list(QString const &prefix)
 {
-    char *date;
     QString signData;
     QString url;
+
     QNetworkReply *reply;
     S3ListResult *result;
 
-    date = s3_make_date();
-    signData = "GET\n\n\n" + QString(date) + "\n/" + m_bucket + "/";
     url = "http://" + m_baseUrl + "/" + m_bucket + "/?delimiter=/";
     if (!prefix.isEmpty()) {
         url += "&prefix=" + prefix;
     }
 
-    reply = performOp(GET, url, signData, date, nullptr, nullptr, nullptr);
+    reply = performOp(GET, url, nullptr, nullptr, nullptr);
     result = new S3ListResult(reply, this);
-    free(date);
 
     return result;
 }
 
-QNetworkReply *S3Access::performOp(Method method, QString const &url, QString const &sign_data, const char *date, QIODevice *in, const char *content_md5, const char *content_type)
+QNetworkReply *S3Access::performOp(Method method, QString const &url, QIODevice *in, const char *content_md5, const char *content_type, QString signDataKey)
 {
     QNetworkRequest request;
-    char *digest;
+    unsigned char digest[EVP_MAX_MD_SIZE +1 ]; // +1 for the \0 byte
+    QByteArray digest_base64;
+    unsigned int out_length = 0;
+    QString methodStr;
+    QString date = QDateTime::currentDateTimeUtc().toString(Qt::RFC2822Date);
     QNetworkReply *reply;
 
+    switch (method) {
+    case DELETE:
+        methodStr = "DELETE";
+        qDebug() << "DELETE request";
+        break;
+    case PUT:
+        methodStr = "PUT";
+        qDebug() << "PUT request";
+        break;
+    case GET:
+    default: // GET
+        methodStr = "GET";
+        qDebug() << "GET request";
+        break;
+    }
+
+    QString signData = methodStr + "\n\n\n" +   + "\n/" + m_bucket + "/";
+
+    if (signDataKey != nullptr)
+        signData+=signDataKey;
```

Comment:
For consistency (and because it's a common source of error) I'd be grateful if you could include the curly brackets here.